Client/Frontend/Settings/Brave Today/BraveTodayAddSourceViewController.swift

```diff
@@ -0,0 +1,371 @@
+                )
+                self.navigationController?.pushViewController(resultsController, animated: true)
+            case .failure(let error):
+                self.displayError(error)
+            }
+        }
+    }
+    
+    private enum FindFeedsError: Error {
+        /// An error occured while attempting to download the page
+        case dataTaskError(Error)
+        /// The data was either not received or is in the incorrect format
+        case invalidData
+        /// The data downloaded did not match a
+        case parserError(ParserError)
+        /// No feeds were found at the given URL
+        case noFeedsFound
+        
+        var localizedDescription: String {
+            switch self {
+            case .dataTaskError(let error as URLError) where error.code == .notConnectedToInternet:
+                return error.localizedDescription
+            case .dataTaskError:
+                return Strings.BraveToday.addSourceNetworkFailureMessage
+            case .invalidData, .parserError:
+                return Strings.BraveToday.addSourceInvalidDataMessage
+            case .noFeedsFound:
+                return Strings.BraveToday.addSourceNoFeedsFoundMessage
+            }
+        }
+    }
+    
+    private var pageTask: URLSessionDataTask?
+    private func downloadPageData(for url: URL, _ completion: @escaping (Result<[RSSFeedLocation], FindFeedsError>) -> Void) {
+        pageTask = session.dataTask(with: url) { [weak self] (data, response, error) in
+            guard let self = self else { return }
+            if let error = error as? URLError, error.code == .cancelled {
+                return
+            }
+            if let error = error {
+                completion(.failure(.dataTaskError(error)))
+                return
+            }
+            guard let data = data, let root = try? HTMLDocument(data: data) else {
+                completion(.failure(.invalidData))
+                return
+            }
+            
+            // Check if `data` is actually an RSS feed
+            if case .success(let feed) = FeedParser(data: data).parse() {
+                // User provided a direct feed
+                var title: String?
+                switch feed {
+                case .atom(let atom):
+                    title = atom.title
+                case .json(let json):
+                    title = json.title
+                case .rss(let rss):
+                    title = rss.title
+                }
+                completion(.success([.init(title: title, url: url)]))
+                return
+            }
+            
+            // Check if `data` is actually an OPML list
+            if let opml = OPMLParser.parse(data: data), !opml.outlines.isEmpty {
+                let locations = opml.outlines.compactMap(self.rssLocationFromOPMLOutline)
+                if !locations.isEmpty {
+                    completion(.success(locations))
+                    return
+                }
```

Comment:
No need to complete failure here if locations are empty?

Reply:
I was going to say it could continue on the flow, but if it successfully parsed as OPML and got here then its probably not going to find any feed information in the HTML, so yeah I can add a failure here